### PR TITLE
Java

### PR DESCRIPTION
**Pull request by yuki-stripe**
### Testing
- [x] 4000 0025 0000 3155 - 3DS1
- [x] 4000 0000 0000 3220 -> shows native 3ds2 flow -> success
- [x] 4000 0084 0000 1629 -> decline.

### CODE REVIEW

File: client/android-java/app/src/test/java/com/example/app/ExampleUnitTest.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test
+
+import org.junit.Assert.*
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
Comment (mshafrir-stripe) [inline client/android-java/app/src/test/java/com/example/app/ExampleUnitTest.kt]: delete

File: client/android-java/app/src/main/res/xml/network_security_config.xml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<network-security-config>
```
Comment (mshafrir-stripe) [inline client/android-java/app/src/main/res/xml/network_security_config.xml:2]: is this for Charles?

File: client/android-java/app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+<resources>
+    <string name="app_name">App</string>
```
Comment (mshafrir-stripe) [inline client/android-java/app/src/main/res/values/strings.xml:2]: maybe set the App name to something more descriptive?

File: client/android-java/app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+apply plugin: 'kotlin-android'
+
+apply plugin: 'kotlin-android-extensions'
+
+android {
+    compileSdkVersion 28
+    defaultConfig {
+        applicationId "com.example.app"
+        minSdkVersion 21
+        targetSdkVersion 28
```
Comment (mshafrir-stripe) [inline client/android-java/app/build.gradle]: you can set this to 29

File: client/android-java/app/build.gradle [hunk 5]
```diff
@@ -0,0 +1,46 @@
+apply plugin: 'com.android.application'
+
+apply plugin: 'kotlin-android'
+
+apply plugin: 'kotlin-android-extensions'
+
+android {
+    compileSdkVersion 28
```
Comment (mshafrir-stripe) [inline client/android-java/app/build.gradle]: you can set this to 29

File: client/android-java/app/build.gradle [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    defaultConfig {
+        applicationId "com.example.app"
```
Comment (mshafrir-stripe) [inline client/android-java/app/build.gradle:10]: let's set a better app id - something like `com.stripe.android.payments`

File: client/android-java/app/build.gradle [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation 'androidx.appcompat:appcompat:1.0.0-beta01'
+    implementation 'androidx.constraintlayout:constraintlayout:1.1.2'
+    implementation 'com.google.android.material:material:1.0.0-beta01'
+    testImplementation 'junit:junit:4.12'
+    androidTestImplementation 'androidx.test:runner:1.1.0-alpha4'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.0-alpha4'
+
+    // Stripe Android SDK
+    implementation 'com.stripe:stripe-android:11.1.3'
```
Comment (mshafrir-stripe) [inline client/android-java/app/build.gradle]: 11.2.0

File: client/android-java/app/build.gradle [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    compileOptions {
+        sourceCompatibility 1.8
+        targetCompatibility 1.8
+    }
+}
+
+dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
+    implementation"org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
```
Comment (mshafrir-stripe) [inline client/android-java/app/build.gradle]: shouldn't need Kotlin here

File: client/android-java/app/build.gradle [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    testImplementation 'junit:junit:4.12'
+    androidTestImplementation 'androidx.test:runner:1.1.0-alpha4'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.0-alpha4'
+
+    // Stripe Android SDK
+    implementation 'com.stripe:stripe-android:11.1.3'
+
+    // OkHttp & GSON
+    implementation 'com.squareup.okhttp3:okhttp:4.2.0'
+    implementation 'com.google.code.gson:gson:2.8.5'
```
Comment (mshafrir-stripe) [inline client/android-java/app/build.gradle]: Let's use latest

```
    implementation fileTree(dir: 'libs', include: ['*.jar'])
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
    implementation 'com.google.android.material:material:1.0.0'
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'androidx.test:runner:1.2.0'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'

    // Stripe Android SDK
    implementation 'com.stripe:stripe-android:11.2.0'

    // OkHttp & GSON
    implementation 'com.squareup.okhttp3:okhttp:4.2.2'
    implementation 'com.google.code.gson:gson:2.8.6'
```

File: client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    /**
+     * To run this app, you'll need to first run the sample server locally.
+     * Follow the "How to run locally" instructions in the root directory's README.md to get started.
+     * Once you've started the server, open http://localhost:4242 in your browser to check that the
+     * server is running locally.
+     * After verifying the sample server is running locally, build and run the app using the
+     * Android emulator.
+     */
+    // 10.0.2.2 is the Android emulator's alias to localhost
+    private String backendUrl = "http://10.0.2.2:4242/";
```
Comment (mshafrir-stripe) [inline client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java]: `private static final String BACKEND_URL = `

File: client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+                        }
+                    });
+                }
+            }
+
+            @Override
+            public void onError(@NonNull Exception e) {
+                // Setup request failed – allow retrying using the same payment method
+                runOnUiThread(() -> {
+                    if (weakActivity.get() != null) {
```
Comment (mshafrir-stripe) [inline client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java]: Set to a local variable.

```
Activity activity = weakActivity.get();
if (activity != null) {
  ...
}
```

File: client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+                                loadPage();
+                            });
+                            AlertDialog dialog = builder.create();
+                            dialog.show();
+                        }
+                    });
+                } else if (status == SetupIntent.Status.RequiresPaymentMethod) {
+                    // Setup failed – allow retrying using a different payment method
+                    runOnUiThread(() -> {
+                        if (weakActivity.get() != null) {
```
Comment (mshafrir-stripe) [inline client/android-java/app/src/main/java/com/example/app/CheckoutActivity.java]: Set to a local variable.

**Review by mshafrir-stripe — CHANGES_REQUESTED** [review]

File: client/android-java/app/build.gradle (hunk 6 above)
Comment (yuki-stripe) [reply client/android-java/app/build.gradle:10]: @bg-stripe thoughts?

I don't feel strongly either way, the project is just a container for the sample code imo.

Comment (mshafrir-stripe) [reply client/android-java/app/build.gradle:10]: I don't feel strongly either, this is fine

File: client/android-java/app/src/main/res/xml/network_security_config.xml (hunk 2 above)
Comment (yuki-stripe) [reply client/android-java/app/src/main/res/xml/network_security_config.xml:2]: IDK, but no need to make this work with Charles.  We do want this to work with a server running locally.

File: client/android-java/app/src/main/res/values/strings.xml (hunk 3 above)
Comment (yuki-stripe) [reply client/android-java/app/src/main/res/values/strings.xml:2]: I'm down to punt on this for now (alllll the samples are currently called app, I'd rather make one decision to keep or do a pass to rename them all).

**Pull request merged**